### PR TITLE
Add DB models & views for fitting specifications

### PR DESCRIPTION
**Pull request by jonc125**
Part of #203.

I've managed to generalise the entity views so most can now be used as-is for fitting specifications, for which the `FittingSpec` DB model has been added. It inherits from `Entity` but unlike the model & protocol cases has its own table in the DB to hold the protocol link.

The only views that have different versions for fitting specs are the following, so they can use different forms:
* FittingSpecCreateView (adds the protocol link)
* FittingSpecNewVersionView (removes the re-run experiments option)

All the following views are re-used for fitting specs at present. In several cases the templates or views have been adapted so some elements are only used for models/protocols.
- EntityListView
- EntityView
- EntityVersionView (the Protocol link is not yet shown)
- EntityVersionJsonView (planned experiments are not checked for fitting specs)
- EntityComparisonView & EntityComparisonJsonView
- EntityTagVersionView
- EntityDeleteView
- EntityVersionListView
- EntityArchiveView
- FileUploadView
- ChangeVisibilityView
- EntityFileDownloadView
- EntityCollaboratorsView
- EntityDiffView
- EntityAlterFileView (not actually used, but trivial to include and saved avoiding linking to it)

The other entity views don't apply to fitting specs, although variations on the first two might when we have fitting results:
- EntityCompareExperimentsView
- EntityRunExperimentView
- CheckProtocolCallbackView
- GetProtocolInterfacesJsonView

To make the reuse work several things were needed:
* `EntityTypeMixin` now knows about `FittingSpec`
* We have `url_type` and `display_type` fields as well as `entity_type` to help with user-friendliness
* A `current_namespace` context item is available in all templates so URL reversing can tell which app we're currently in (fitting or entities)
    * Added to `core/context_processors.py`. cf e.g. https://stackoverflow.com/questions/20020911/how-to-get-the-name-of-current-app-within-a-template
    * If `self` is a view, use `self.request.resolver_match.namespace` to get at it.
* Changed entity tags significantly to work with the above

Other bits of note:
* Created `EntityManager.visible_to_user` query for filtering protocols that can be linked to a new fitting spec. Also improved `shared_with_user` so it can be combined without using `.union()`. This will be handy elsewhere too: at least when creating datasets, and possibly in matrix views.

For future consideration:
* The JSON views that serve up version & file info are almost-duplicated in lots of places. It might be best to move e.g. the bulk of EntityVersionJsonView into a mixin, and let subclasses override the url reversing.
  It's similar to DatasetJsonView & ExperimentVersionJsonView (though not quite identical), and parts of ExperimentComparisonJsonView.
  There's also overlap with EntityComparisonJsonView.

### CODE REVIEW

**Comment by jonc125** [comment]
With 9f77284c I'm thinking about how to implement views for fitting specs, initially around creation/upload. I don't want to have to duplicate lots of code and templates from the entities app. But equally some things will be different (like the new spec form having a protocol link) so some overrides will be necessary. I think doing it neatly will need some alterations to the generic entities views; I'm still thinking about how best to do so. Some will just be stylistic, e.g. adding a 'fitting specs' button alongside models and protocols; making the 'Your fittingspecs' heading decent English!

If running this locally you can see what I've implemented this far at http://localhost:8000/fitting/fittingspecs/

**Comment by jonc125** [comment]
This needs unit tests to be written for the fitting spec views, but from manual button pressing everything seems to work as expected. So probably worth reviewing and merging, with testing in a separate PR @sroderick-g5sro ? I've updated the description so it explains what was done.

**Comment by jonc125** [comment]
Note before this is merged: we'll need to merge the latest master in to this branch then do `manage.py makemigrations`, since the fitting spec cache will need to have the new fields added.

**Comment by sroderick-g5sro** [comment]
I've started my review of this, but its going to take me sometime to understand the background of this change and then work through it. I am keen to try and use if from a user's perspective.

**Comment by jonc125** [comment]
> I've started my review of this, but its going to take me sometime to understand the background of this change and then work through it. I am keen to try and use if from a user's perspective.

That's expected! I have no meetings this morning, so could do a call to walk you through the user's POV. It's discussed in text form in #135. For this PR, basically it's treating fitting specifications like models & protocols, in that they're versioned collections of files that a user can upload, backed by a git repository. So the same creation/view/etc UI applies. But with slight tweaks because they link to a protocol.

Which reminds me - I don't think I've included that protocol link in the detail view yet! Will try to add that today...

**Comment by sroderick-g5sro** [comment]
Just a trivial point that when user is deleting an entity or fitting spec. Get a confirm button but not a cancel one. So have to use browser back. Can't remember if it was always like this - assume so

**Comment by sroderick-g5sro** [comment]
Sorry about v. minor point - when clicking on model / protocol /etc. you get some "flash of unformatted comment" I think this is becuase takes time for the version number and detail to show up. probably not worth worrying about.

**Comment by sroderick-g5sro** [comment]
I'm not too keen on the way that the "Compare..." button works. It took me a while to figure out that it contained a clickable link.

**Comment by jonc125** [comment]
> Just a trivial point that when user is deleting an entity or fitting spec. Get a confirm button but not a cancel one. So have to use browser back. Can't remember if it was always like this - assume so

Yes, I think that was always the case. A cancel button as well makes sense though!

> Sorry about v. minor point - when clicking on model / protocol /etc. you get some "flash of unformatted content"

Yes, that's in the delay while the Javascript code fetches the JSON metadata for the fitting spec and uses it to fill in the remaining page content. We could make the interim page a bit nicer though!

> I'm not too keen on the way that the "Compare..." button works. It took me a while to figure out that it contained a clickable link.

For models & protocols there are at least two links there, which is why it behaves like a drop-down menu on hover. There may be a better way to do it, but that was the best I came up with, without using up too much screen space!

**Comment by sroderick-g5sro** [comment]
From the fitting specifications "tab" clicking on the name and "add new version" go to the same location. I expected to be presented with a list of files when clicking on the name. (I felt at one point it was doing this). 
[2020-01-17_14-37-33.zip](https://github.com/ModellingWebLab/WebLab/files/4077269/2020-01-17_14-37-33.zip)
Please see video clip for replication. This is not the same behaviour as for models / protocols / datasets.

**Comment by jonc125** [comment]
That should happen if the original creation process was interrupted, i.e. if the spec was created in the DB but no files were added at the time. (It also happens for models and protocols I believe.)

**Comment by sroderick-g5sro** [comment]
I get error building site when trying to compare to versions of a fitting spec

image

**Comment by jonc125** [comment]
I saw that, but thought I'd fixed `compare.js` so it doesn't happen. Have you re-built the static resources with `gulp`? I tend to keep `gulp watch` running in the background so it happens automatically. See the 'build statics' section of the readme.

**Comment by sroderick-g5sro** [comment]
I have tried refreshing gulp and same is still happening. I have also put a couple of minor questions in the review

**Comment by jonc125** [comment]
That's odd, it's working locally. Maybe refresh the page just in case your browser has cached the JS?

**Comment by jonc125** [comment]
I have noticed that the 'click to show single version' feature in the drop-down isn't working once you get to the page though. But then, it's not working for models or protocols either!

**Comment by sroderick-g5sro** [comment]
> That's odd, it's working locally. Maybe refresh the page just in case your browser has cached the JS?

yes sorry was a caching problem - thanks

File: weblab/entities/views.py [hunk 1]
```diff
@@ -62,16 +63,26 @@ class EntityTypeMixin:
...
+            if et.url_type == self.kwargs['entity_type']
+        )
+
+    @property
+    def other_model(self):
         return next(
             et
             for et in (ModelEntity, ProtocolEntity)
-            if et.entity_type == self.kwargs['entity_type']
+            if et.other_type == self.kwargs['entity_type']
```
[the remark this answers is not in the document]
Comment (jonc125) [reply weblab/entities/views.py]: It is, yes. The concept doesn't make sense for `FittingSpec` - it's designed to be the other half in 'run this protocol on this model' logic.

File: weblab/fitting/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    assert other_user not in fittingspec.collaborators
+    assert not fittingspec.is_editable_by(other_user)
+    assert not fittingspec.is_deletable_by(other_user)
+
+
+@pytest.mark.django_db
+class TestRepository:
+    def test_repo_path_create_and_delete(self, fake_repo_path):
+        spec = recipes.fittingspec.make()
+        path = '%s/%d/fittingspecs/%d' % (fake_repo_path, spec.author.pk, spec.pk)
```
[the remark this answers is not in the document]
Comment (jonc125) [reply weblab/fitting/tests/test_models.py]: Good catch! Will fix.

**Pull request merged**